ports/rp2/machine_pin.c

```diff
@@ -313,51 +313,62 @@ STATIC mp_obj_t machine_pin_toggle(mp_obj_t self_in) {
+        irq->base.handler = mp_const_none;
+        irq->base.ishard = false;
+        MP_STATE_PORT(machine_pin_irq_obj[pin]) = irq;
+    }
+
+    // Disable all IRQs while data is updated.
+    gpio_set_irq_enabled(pin, GPIO_IRQ_ALL, false);
+
+    // Update IRQ data.
+    irq->base.handler = handler;
+    irq->base.ishard = hard;
+    irq->flags = 0;
+    irq->trigger = trigger;
+
+    // Enable IRQ if a handler is given.
+    if (handler != mp_const_none && trigger != MP_HAL_PIN_TRIGGER_NONE) {
+        bool enabled = trigger != MP_HAL_PIN_TRIGGER_NONE;
+        gpio_set_irq_enabled(pin, trigger, enabled);
+    }
+}
+
 // pin.irq(handler=None, trigger=IRQ_FALLING|IRQ_RISING, hard=False)
 STATIC mp_obj_t machine_pin_irq(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
     enum { ARG_handler, ARG_trigger, ARG_hard };
     static const mp_arg_t allowed_args[] = {
         { MP_QSTR_handler, MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} },
-        { MP_QSTR_trigger, MP_ARG_INT, {.u_int = GPIO_IRQ_EDGE_FALL | GPIO_IRQ_EDGE_RISE} },
+        { MP_QSTR_trigger, MP_ARG_INT, {.u_int = MP_HAL_PIN_TRIGGER_FALL | MP_HAL_PIN_TRIGGER_RISE} },
         { MP_QSTR_hard, MP_ARG_BOOL, {.u_bool = false} },
     };
     machine_pin_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
     mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
     mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
 
-    // Get the IRQ object.
-    machine_pin_irq_obj_t *irq = MP_STATE_PORT(machine_pin_irq_obj[self->id]);
 
-    // Allocate the IRQ object if it doesn't already exist.
-    if (irq == NULL) {
-        irq = m_new_obj(machine_pin_irq_obj_t);
-        irq->base.base.type = &mp_irq_type;
-        irq->base.methods = (mp_irq_methods_t *)&machine_pin_irq_methods;
-        irq->base.parent = MP_OBJ_FROM_PTR(self);
-        irq->base.handler = mp_const_none;
-        irq->base.ishard = false;
-        MP_STATE_PORT(machine_pin_irq_obj[self->id]) = irq;
-    }
+    mp_obj_t handler = NULL;
+    mp_uint_t trigger = MP_HAL_PIN_TRIGGER_NONE;
+    bool hard = false;
 
     if (n_args > 1 || kw_args->used != 0) {
-        // Configure IRQ.
-
-        // Disable all IRQs while data is updated.
-        gpio_set_irq_enabled(self->id, GPIO_IRQ_ALL, false);
-
         // Update IRQ data.
-        irq->base.handler = args[ARG_handler].u_obj;
-        irq->base.ishard = args[ARG_hard].u_bool;
-        irq->flags = 0;
-        irq->trigger = args[ARG_trigger].u_int;
-
-        // Enable IRQ if a handler is given.
-        if (args[ARG_handler].u_obj != mp_const_none) {
-            gpio_set_irq_enabled(self->id, args[ARG_trigger].u_int, true);
-        }
+        handler = args[ARG_handler].u_obj;
+        hard = args[ARG_hard].u_bool;
+        trigger = args[ARG_trigger].u_int;
     }
-
-    return MP_OBJ_FROM_PTR(irq);
+    machine_pin_irq_helper(self->id, handler, trigger, hard);
```

Comment:
Doing it this way changes semantics: if the user does `pin.irq()` to retrieve the current IRQ object then it will now disable the callback (previously it would just leave the IRQ object alone).

Reply:
Great catch, I did indeed miss that usage. I've split this up slightly to better allow maintaining the current usage.